### ISSUE TITLE
[💡SUG] Get a prediction

### ISSUE THREAD

**Issue by mayaKaplansky** [RUCAIBox/RecBole#632]
Hi
How can I get a single prediction from a sequential model?
I have the *.pth file, how can I use it?

**Comment by 2017pxy** [comment]
@mayaKaplansky Firstly, about `.pth` file. You can use the following code to load the model.
```
 checkpoint = torch.load('.pth')
model.load_state_dict(checkpoint['state_dict'])
 ```
And then, about getting a single prediction from a sequential model. Firstly you need to construct a `interaction`, you can click [here](https://recbole.io/docs/user_guide/data/interaction.html?highlight=interaction) for more details about `interaction`.

**Comment by mayaKaplansky** [comment]
Hi Thanks
I am looking for a way to use `model.full_sort_predict(interaction)` to give as a parameter a list of items and get the score of the prediction of the items for a sequential model. I see that I can give a user which is a session in sequential model so it doesn't make sense, as each session happens only once so the prediction is known. In sequential model I am supposed to predict the next item based on the items already seen. Is there a way to get prediction for such a use case?
thanks

**Comment by chenyushuo** [comment]
Here is a simple usage of `model.full_sort_predict(interaction)`. I wish it could help you to get the prediction of sequential model.

```python3
import numpy as np
import torch
from recbole.model.sequential_recommender.gru4rec import GRU4Rec
from recbole.config import Config
from recbole.data import create_dataset, data_preparation
from recbole.data.interaction import Interaction


if __name__ == '__main__':
    # this part is to load saved model.
    config_dict = {
        # here you can set some parameters such as `gpu_id` and so on.
    }
    config = Config(model='GRU4Rec', dataset='ml-100k', config_dict=config_dict)
    dataset = create_dataset(config)
    train_data, valid_data, test_data = data_preparation(config, dataset)
    model = GRU4Rec(config, train_data)
    checkpoint = torch.load('RecBole/saved/GRU4Rec-Dec-08-2020_15-37-37.pth')  # *.pth file is used for load saved model.
    model.load_state_dict(checkpoint['state_dict'])
    model.eval()

    # Here is a example to predict user 1's and user 2's next item.
    # user 1 have seen item 1, item 2, item 3
    # user 2 have seen item 4, item 5
    # 0 is used for padding.
    # scores is a [2, n_items] matrix which represent the score of every items.
    input_inter = Interaction({
        'user_id': torch.tensor([1, 2]),
        'item_id_list': torch.tensor([[1, 2, 3, 0, 0],
                                      [4, 5, 0, 0, 0]]),
        'item_length': torch.tensor([3, 2]),
    })
    with torch.no_grad():
        scores = model.full_sort_predict(input_inter)
    print(scores.shape)
    print(scores)
```

![screen shot](https://user-images.githubusercontent.com/31039063/104116263-eab32f00-5351-11eb-8c19-387e820779d5.png)

**Comment by mayaKaplansky** [comment]
Thank you for your help!
Is '`user id`' new IDs I can take from my test data and not existing users from train?
The reason I'm asking, is that in sequential model, the instructions are that the` user_id` is actually the` session_id.` Therefore, each session happens only once and its known - so there is nothing to predict. The idea is to predict what happens in a new session.
I would expect that as an input for the 'Interaction' I would just enter the` item_id_list`. Can you help me understand that?

Also what does the size of `'item_length': torch.tensor([3, 2])`, is 3 by 2? What does this size represent?
Many thanks! I am really anxious to succeed using this library for my article :)

**Comment by batmanfly** [comment]
@mayaKaplansky Hi, thank you for using our library for your research. Let me try to reply to your query (instead of the implementation based on RecBole). 

(1) For the first query, I think there are two different settings for session-based recommendation. The first one is to predict the entire content (all the items) of a new session given historical sessions; the second one is to predict the last item of a new session also given historical sessions. (based on my understanding, many session-based algorithms also assume an implicit user ID would exist who would create a sequence of sessions, but some don't) The setting would be determined by your own task. For the implementation in RecBole, I believe we have incorporated several session-based recommenders (you can also see our data copy for session-based recommendation). You may read and run the code for your understanding. 

(2) For the second query, could you please run the code and simply check this?

I suggest you create toy-example datasets in your format (e.g., three to four users, and three to four items) to understand session-based algorithms implemented in our RecBole.

**Comment by mayaKaplansky** [comment]
Hi
according to this thread: https://github.com/RUCAIBox/RecBole/issues/617 , in sequential models we replace the concept of `user_id` into the `session_id`, which will appear only once in the data. For me, this contradicts this statement: 

> many session-based algorithms also assume an implicit user ID would exist who would create a sequence of sessions, but some don't

Since according to the thread,  I am not providing any user ID. 

In addition ,according to the website ,in the sequential models, the `full_sort_predict` is defined like this: 
![image](https://user-images.githubusercontent.com/57919041/104238733-55975f80-5451-11eb-8a1a-d2f398a163c1.png)

this means that the input that is needed is user ID which is not part of sequential, and also the list of seen items does not affect the calculation. It sounds like the function calculates the chances a user will click on a specific item. But what about the seen items in the session?

I am searching deeply in the website, and the code (as much as I can) and this forum before I post a question.

All I'm trying to do is to get a series of predictions based on given list of clicks, where I already have my *.pth file. 

Many many thanks - it would be a great achievement for me to include this library in my research article.

**Comment by batmanfly** [comment]
@mayaKaplansky 

[For clarifying your query] Just confirm your queries. So far, you seem to be focused on the question "each session appears only once, how do you predict the future session since the session ID would be new"? That question was raised in a setting "a user corresponds to multiple sessions, but there is no user ID at all. Your task is to predict the items of the next session".
Your query is originally stated as "get a series of predictions based on given list of clicks". The reason why I asked this is I need to confirm your task setting and its difference with previous session-based recommendation algorithms. 

[For understanding session-based recommendation] If so, the setting is exactly same as that in original GRU4Rec paper (I'm not in charge of its implementation. So I just commented based on the paper).  In this way, I would strongly suggest you read the paper of GRU4Rec paper. I believe you would get the understanding of session-based recommendation.  

[For understanding our implementation about GRU4Rec] The persons in my teams who are implementing the code of GRU4Rec and other session-based algorithms would also reply based on the implementation in our library.

**Comment by lyj1998** [comment]
> Hi
> according to this thread: #617 , in sequential models we replace the concept of `user_id` into the `session_id`, which will appear only once in the data. For me, this contradicts this statement:
> 
> > many session-based algorithms also assume an implicit user ID would exist who would create a sequence of sessions, but some don't
> 
> Since according to the thread, I am not providing any user ID.
> 
> In addition ,according to the website ,in the sequential models, the `full_sort_predict` is defined like this:
> ![image](https://user-images.githubusercontent.com/57919041/104238733-55975f80-5451-11eb-8a1a-d2f398a163c1.png)
> 
> this means that the input that is needed is user ID which is not part of sequential, and also the list of seen items does not affect the calculation. It sounds like the function calculates the chances a user will click on a specific item. But what about the seen items in the session?
> 
> I am searching deeply in the website, and the code (as much as I can) and this forum before I post a question.
> 
> All I'm trying to do is to get a series of predictions based on given list of clicks, where I already have my *.pth file.
> 
> Many many thanks - it would be a great achievement for me to include this library in my research article.






Session-based recommendation aims to predict which item a user would like to click next, only based upon his/her current interaction sequence.For instance, the problem can be stated as<a href="https://www.codecogs.com/eqnedit.php?latex=[i_4,i_2,i_3,?]" target="_blank"><img src="https://latex.codecogs.com/gif.latex?[i_4,i_2,i_3,?]" title="[i_4,i_2,i_3,?]" /></a>.<a href="https://www.codecogs.com/eqnedit.php?latex=[i_4,i_2,i_3]" target="_blank"><img src="https://latex.codecogs.com/gif.latex?[i_4,i_2,i_3]" title="[i_4,i_2,i_3]" /></a> as input to the model, we predict the next item. Then I will demonstrate our implementation of GRU4Rec to help you understand better.
[our implementation about GRU4Rec] :

- **The input of model**:The click sequence in the session, <a href="https://www.codecogs.com/eqnedit.php?latex=x=[x_1,x_2,...,x_r],1\leq&space;r\leq&space;n" target="_blank"><img src="https://latex.codecogs.com/gif.latex?x=[x_1,x_2,...,x_r],1\leq&space;r\leq&space;n" title="x=[x_1,x_2,...,x_r],1\leq r\leq n" /></a>, is encoded by one hot encoding and compressed into a low-dimensional continuous vector by the embedding layer as the input of GRU.

- **The output of model**:The predicted probability of each candidate item being clicked,<a href="https://www.codecogs.com/eqnedit.php?latex=y&space;=M(x)" target="_blank"><img src="https://latex.codecogs.com/gif.latex?y&space;=M(x)" title="y =M(x)" /></a> , where <a href="https://www.codecogs.com/eqnedit.php?latex=y=[y_1,y_2,y_3,..,y_m]" target="_blank"><img src="https://latex.codecogs.com/gif.latex?y=[y_1,y_2,y_3,..,y_m]" title="y=[y_1,y_2,y_3,..,y_m]" /></a>
![image](https://user-images.githubusercontent.com/35167256/104268420-d6487100-54ce-11eb-94dc-dbea2b0b2f94.png)

We combine code to understand：
```
class GRU4Rec(SequentialRecommender):
    def __init__(self, config, dataset):
        super(GRU4Rec, self).__init__(config, dataset)
        #1. load parameters info, you can customize the yaml file for settings
        self.embedding_size = config['embedding_size']
        self.hidden_size = config['hidden_size']
        self.loss_type = config['loss_type']
        self.num_layers = config['num_layers']
        self.dropout_prob = config['dropout_prob']

        # 2.define layers and loss
# Embedding Layer
        self.item_embedding = nn.Embedding(self.n_items, self.embedding_size, padding_idx=0)
# Dropout Layer       
 self.emb_dropout = nn.Dropout(self.dropout_prob)
# GRU Layer        
self.gru_layers = nn.GRU(
            input_size=self.embedding_size,
            hidden_size=self.hidden_size,
            num_layers=self.num_layers,
            bias=False,
            batch_first=True,
        )
        self.dense = nn.Linear(self.hidden_size, self.embedding_size)

# Here is a example.
# batch_size is set to 2, MAX_ITEM_LIST_LENGTH is set to 5, embedding_size is set to 5
    # the first session sequences: item 1, item 2, item 3
    # the second session sequences: item 4, item 5
    # 0 is used for padding.
    # seq_output is a matrix which represents the embedding vector of predicted item in each sequence.
# item_seq :torch.tensor([[1, 2, 3, 0, 0], [4, 5, 0, 0, 0]])
# item_length: torch.tensor([3, 2]),

# seq_output：torch.tensor([[ 1.0768,  1.3084,  0.6908,  0.2425, -0.8682],
                         [ 1.5825,  0.3526,  0.7171, -0.7511, -0.0954]]
    def forward(self, item_seq, item_seq_len):
        item_seq_emb = self.item_embedding(item_seq)
        item_seq_emb_dropout = self.emb_dropout(item_seq_emb)
        gru_output, _ = self.gru_layers(item_seq_emb_dropout)
        gru_output = self.dense(gru_output)
        # the embedding of the predicted item, shape of (batch_size, embedding_size)
        seq_output = self.gather_indexes(gru_output, item_seq_len - 1)
        return seq_output
```

**Comment by mayaKaplansky** [comment]
Thank you all for your great explanations!
I played with it, and it seems that in the sequential model, the `user_id` does not effect the scores since its just some session ID.

anyhow, in my data I have added features on the inter file, that represent characteristics of the session, which should affect the model.
I stated this in my yaml file:
```
load_col:
    inter: [session_id, item_id, timestamp, PatientLocationID,GenderID,AgeGroup, JobGroup]

```
and saw that in the dataset indeed these columns are converted to lists.

my question is: is it possible to add the features as input for prediction? Same as I provide the list of items already seen, I would like to produce the features so a better prediction will be made (assuming they were using on training)?

Many thanks!

**Comment by batmanfly** [comment]
@mayaKaplansky hi,  you can refer to the implementation of GRU4RecF and GRU4RecKG 

https://github.com/RUCAIBox/RecBole/blob/master/recbole/model/sequential_recommender/gru4reckg.py

https://github.com/RUCAIBox/RecBole/blob/master/recbole/model/sequential_recommender/gru4recf.py

**Comment by mayaKaplansky** [comment]
Thank you!
according to this: https://github.com/RUCAIBox/RecBole/issues/608
I am using GRU4Rec, as GRU4RecF is meant for features for items and features for users, whereas my features are on the interaction level. GRU4RecKG is meant for a graph between items which also is not my use case.
So, is it still correct that I should have used GRU4Rec?

If so , I had to also change the code to include the features like this:

```
class SequentialRecommender(AbstractRecommender):
    """
    This is a abstract sequential recommender. All the sequential model should implement This class.
    """
    type = ModelType.SEQUENTIAL

    def __init__(self, config, dataset):
        super(SequentialRecommender, self).__init__()

        # load dataset info
        self.USER_ID = config['USER_ID_FIELD']
        self.ITEM_ID = config['ITEM_ID_FIELD']
        self.ITEM_SEQ = self.ITEM_ID + config['LIST_SUFFIX']
        self.ITEM_SEQ_LEN = config['ITEM_LIST_LENGTH_FIELD']
        self.POS_ITEM_ID = self.ITEM_ID
        self.NEG_ITEM_ID = config['NEG_PREFIX'] + self.ITEM_ID
        self.max_seq_length = config['MAX_ITEM_LIST_LENGTH']
        self.n_items = dataset.num(self.ITEM_ID)
        self.JobGroup = config['JobGroup_FIELD']
        self.JobGroup_SEQ = self.JobGroup + config['LIST_SUFFIX']
        self.AgeGroup = config['AgeGroup_FIELD']
        self.AgeGroup_SEQ = self.AgeGroup + config['LIST_SUFFIX']
        self.GenderID = config['GenderID_FIELD']
        self.GenderID_SEQ = self.GenderID + config['LIST_SUFFIX']
        self.PatientLocationID = config['PatientLocationID_FIELD']
        self.PatientLocationID = self.PatientLocationID + config['LIST_SUFFIX']

```
Also, I need to update the loss function.

After training, I would like to predict the next Item ID based on previous clicks and the given features.
But in the 
```
    def full_sort_predict(self, interaction):
        item_seq = interaction[self.ITEM_SEQ]
        item_seq_len = interaction[self.ITEM_SEQ_LEN]
        seq_output = self.forward(item_seq, item_seq_len)
        test_items_emb = self.item_embedding.weight
        scores = torch.matmul(seq_output, test_items_emb.transpose(0, 1))  # [B n_items]
        return scores
```
I see I cannot provide the features as input to be used. Is there a way to add the features as parameters for predictions?

**Comment by aHuiWang** [comment]
@mayaKaplansky 
1. I think you mentioned 'interaction level' features are exactly user's feature (i.e. session's feature) or item's feature. So you should put the user's features (e.g. age and gender) in the *.user file and put the item's features (e.g. category and color) in the *.item file. And you also should modify the dataset configuration file to load features.
```
load_col:
    inter: [session_id, item_id, timestamp]
    item: [item_id, category, color]
    user: [session_id, age, gender]  
```
2. As for the use of item's feature, you could refer to the GRU4RecF. The user's feature can be used in a similar way.
3. In our implementation, we maintain the user's feature table and item's feature table and we will get the feature according to the item_seq in the forward function. 
[code](https://github.com/RUCAIBox/RecBole/blob/ae5c5415c8453e68ba17a5e53afdbb6f0e9f550b/recbole/model/sequential_recommender/gru4recf.py#L94-L96)

**Comment by batmanfly** [comment]
@mayaKaplansky  Just add one point. For features from user side, unlike item features, it may not be useful to be involved in intermediate process (we concatenate item embedding with item feature embedding, however, we cann't typically concatenate item embedding with user feature embedding ). 

A solution may be you first get a session representation, and then concatenate it with  user feature embedding for final prediction. However, it depends on your problem and model architecture.

**Comment by mayaKaplansky** [comment]
Hi
I am actually referring to interaction features, and not user nor item features. That's why GRU4recf does not fit, and it sounds like I should use GRU4rec.
please see here: https://github.com/RUCAIBox/RecBole/issues/608 I was instructed to add them in the *.inter file, same as you demonstrate the 'rating' column in your documentation.
So, my question still stands, how can I add interaction features to the `full_sort_predict` function?

**Comment by batmanfly** [comment]
@mayaKaplansky I think I finally understood your query. =( 

It is whether you can use interaction features for sequential recommendation or session-based recommendation, right?

If so, we have two kinds of solutions:

(1) Convert the interaction features as user features (session features in your case) or item features. 

Basically, you list features as "    inter: [session_id, item_id, timestamp, PatientLocationID,GenderID,AgeGroup, JobGroup]". What I can see is that "PatientLocationID,GenderID,AgeGroup, JobGroup" belong to the category of user features. If you can specify them as users or items features, the case would become the above response refers to. For example, timestamp refers to the time when a session starts or ends, so you can consider it as a session feature (like user features). If you have multiple users involving in the session, you can also consider the user of user IDs as session features. The case would become similar to what SVDFeature solves. 

In this way, you don’t read user or item features through interaction (solution 2 reads interaction features through interaction)

In this case, you have  gru4recf for reference. 

(2) Directly use interaction features. 

If you feel solution 1 does not work, you should use the following method:

First step: I just copy one reply "Say if you wanna fetch the field named NUM_OF_TIMES, then you need to 1) register the filed name in the config 2) assign an attribute within your customized model, e.g self.NUM_OF_TIMES = config['NUM_OF_TIMES_FIELD'] 3) fetch the field via interaction[self.NUM_OF_TIMES]" from #608

Second step, for your query, "I add interaction features to the full_sort_predict function", after you fetch corresponding interaction features, you can directly use them via the full_sort_predict function or its associated functions forward() and calcuate_loss(). 

But, in this case, we have no example code for you. Because how to use these interaction features (e.g., ) for prediction depends on your task and your network architecture.
  
Let me know whether it is your query and whether it would solve your problem.

**Comment by ShanleiMu** [comment]
@mayaKaplansky I think I finally understood your query. =(

It is whether you can use interaction features for sequential recommendation or session-based recommendation, right?

You list features as " inter: [session_id, item_id, timestamp, PatientLocationID,GenderID,AgeGroup, JobGroup]".

If so, we have two kinds of solutions:

(1) Convert the interaction features as user features (session features in your case) or item features.

Basically, If you can specify them as users or items features, the case would become the above response refers to. For example, timestamp refers to the time when a session starts or ends, so you can consider it as a session feature (like user features). If you have multiple users involving in the session, you can also consider the set of user IDs or their demographic features as session features. The case would become similar to what SVDFeature solves.

In this way, you don’t read user or item features through interaction (solution 2 reads interaction features through interaction)

In this case, you have gru4recf for reference.

(2) Directly use interaction features.

If you feel solution 1 does not work, you should use the following method:

First step: I just copy one nice reply from our contributor "Say if you wanna fetch the field named NUM_OF_TIMES, then you need to 1) register the filed name in the config 2) assign an attribute within your customized model, e.g self.NUM_OF_TIMES = config['NUM_OF_TIMES_FIELD'] 3) fetch the field via interaction[self.NUM_OF_TIMES]" from #608

Second step, for your query, "I add interaction features to the full_sort_predict function", after you fetch corresponding interaction features, you can directly use them via `full_sort_prediction`, `calcualte_loss` for training and predicting, respectively. As for how to use these interaction features for training or prediction, we have no example code for you. Because it depends on your task and your network architecture. Maybe, you can refer to `FeatureSeqEmbLayer` used in GRU4RecF to design your own structures.

```
    def calcualte_loss(self, interaction):
        age_seq = interaction[self.AgeGroup_SEQ]	# (batch_size, seq_length)
        age = interaction[self.AgeGroup]		# (batch_size)
        # and other interaction features
        ...

        # use your own network structures
        ...

        return loss

    def full_sort_predict(self, interaction):
        age_seq = interaction[self.AgeGroup_SEQ]	# (batch_size, seq_length)
        age = interaction[self.AgeGroup]		# (batch_size)
        # and other interaction features
        ...

        # use your own network structures
        ...

        return scores

```

Let me know whether it is your query and whether it would solve your problem.

**Comment by mayaKaplansky** [comment]
Thank you very much. If I would like to try #1, this means that the training data should know about all sessions exist. I mean, I cannot send to the `full_sort_predict` a session ID that was not part of the original data, since it will not know its features. Am I correct in that assumption?

**Comment by batmanfly** [comment]
@mayaKaplansky For this query, I suggest you read the code of GRU4RecF and previous responses.  We have made it clear that we only consider the case for predicting the last item of an existing session (from lyj1998 in this same thread ). In this case, you would have all the information for prediction. This is a commonly adopted evaluation way for session-based recommendation.

**Comment by mayaKaplansky** [comment]
Understood. Thank you!

**Comment by mayaKaplansky** [comment]
> Here is a simple usage of `model.full_sort_predict(interaction)`. I wish it could help you to get the prediction of sequential model.
> 
> ```python
> import numpy as np
> import torch
> from recbole.model.sequential_recommender.gru4rec import GRU4Rec
> from recbole.config import Config
> from recbole.data import create_dataset, data_preparation
> from recbole.data.interaction import Interaction
> 
> 
> if __name__ == '__main__':
>     # this part is to load saved model.
>     config_dict = {
>         # here you can set some parameters such as `gpu_id` and so on.
>     }
>     config = Config(model='GRU4Rec', dataset='ml-100k', config_dict=config_dict)
>     dataset = create_dataset(config)
>     train_data, valid_data, test_data = data_preparation(config, dataset)
>     model = GRU4Rec(config, train_data)
>     checkpoint = torch.load('RecBole/saved/GRU4Rec-Dec-08-2020_15-37-37.pth')  # *.pth file is used for load saved model.
>     model.load_state_dict(checkpoint['state_dict'])
>     model.eval()
> 
>     # Here is a example to predict user 1's and user 2's next item.
>     # user 1 have seen item 1, item 2, item 3
>     # user 2 have seen item 4, item 5
>     # 0 is used for padding.
>     # scores is a [2, n_items] matrix which represent the score of every items.
>     input_inter = Interaction({
>         'user_id': torch.tensor([1, 2]),
>         'item_id_list': torch.tensor([[1, 2, 3, 0, 0],
>                                       [4, 5, 0, 0, 0]]),
>         'item_length': torch.tensor([3, 2]),
>     })
>     with torch.no_grad():
>         scores = model.full_sort_predict(input_inter)
>     print(scores.shape)
>     print(scores)
> ```
> 
> ![screen shot](https://user-images.githubusercontent.com/31039063/104116263-eab32f00-5351-11eb-8c19-387e820779d5.png)

Hi
I have been using this code suggested to me, and I see that only the first itemID in` item_id_list` changes the prediction. Any other input gives me the exact same prediction no matter what.
For example these 2 inputs would give me the exact same prediction: 
```
input_inter = Interaction({
        'user_id': torch.tensor([222]),
        'item_id_list': torch.tensor([[5, 0, 0, 0, 0,0,0,0,0,0]
                                   ]),
        'item_length': torch.tensor([1]),
    })
```

and

```
input_inter = Interaction({
        'user_id': torch.tensor([222]),
        'item_id_list': torch.tensor([[5, 6, 7, 8, 0,0,0,0,0,0]
                                   ]),
        'item_length': torch.tensor([1]),
    })
```
Am I using this wrongly? I expect to get different predictions per input, I tried with hundreds of examples, where according to the metrics you provide, all of them have high scores.
Many thanks

**Comment by mayaKaplansky** [comment]
Just to close the loop here, according to https://github.com/RUCAIBox/RecBole/issues/693 there is no way to load user features and use them in GRU4recF - correct?
